File: server/src/services/recite_plan_service.ts
import { PrismaClient } from '@prisma/client';
import { BusinessError } from '@/middlewares/error_middleware';

const prisma = new PrismaClient();

const EBBINGHAUS = [1, 2, 4, 7, 15, 30, 60];

interface PlanItem {
    id: number;
    userId: number;
    quoteId?: number;
    customQuoteId?: number;
    currentStage: number;
    nextReviewAt: string;
    lastReviewAt?: string;
    masterLevel: number;
    createdAt: string;
}

async function list(userId: number): Promise<PlanItem[]> {
    const plans = await prisma.recitePlan.findMany({
        where: { userId },
        orderBy: { nextReviewAt: 'asc' },
    });

    return plans.map(toPlanItem);
}

async function add(
    userId: number,
    input: { quoteId?: number; customQuoteId?: number }
): Promise<PlanItem> {
    if (!input.quoteId && !input.customQuoteId) {
        throw new BusinessError(400, 'quoteId 或 customQuoteId 不能同时为空');
    }

    const existing = await prisma.recitePlan.findFirst({
        where: {
            userId,
            quoteId: input.quoteId ?? undefined,
            customQuoteId: input.customQuoteId ?? undefined,
        },
    });

    if (existing) {
        return toPlanItem(existing);
    }

    const nextReviewAt = new Date();
    nextReviewAt.setDate(nextReviewAt.getDate() + EBBINGHAUS[0]);

    const plan = await prisma.recitePlan.create({
        data: {
            userId,
            quoteId: input.quoteId,
            customQuoteId: input.customQuoteId,
            currentStage: 0,
            nextReviewAt,
        },
    });

    return toPlanItem(plan);
}

async function todayReview(userId: number): Promise<PlanItem[]> {
    const now = new Date();
    const plans = await prisma.recitePlan.findMany({
        where: { userId, nextReviewAt: { lte: now } },
        orderBy: { nextReviewAt: 'asc' },
    });

    return plans.map(toPlanItem);
}

async function review(
    planId: number,
    userId: number,
    action: 'master' | 'not_skilled' | 'skip'
): Promise<PlanItem> {
    const plan = await prisma.recitePlan.findUnique({ where: { id: planId } });
    if (!plan) throw new BusinessError(404, '背诵计划不存在');
    if (plan.userId !== userId) throw new BusinessError(403, '无权操作');

    let newStage = plan.currentStage;
    let masterLevel = plan.masterLevel;

    if (action === 'master') {
        newStage = Math.min(plan.currentStage + 1, EBBINGHAUS.length - 1);
        masterLevel = Math.min(masterLevel + 1, EBBINGHAUS.length);
    } else if (action === 'not_skilled') {
        newStage = Math.max(plan.currentStage - 1, 0);
    }

    const now = new Date();
    const days = EBBINGHAUS[newStage] ?? EBBINGHAUS[EBBINGHAUS.length - 1];
    const nextReviewAt = new Date(now);
    nextReviewAt.setDate(nextReviewAt.getDate() + days);

    const [updated] = await prisma.$transaction([
        prisma.recitePlan.update({
            where: { id: planId },
            data: {
                currentStage: newStage,
                lastReviewAt: now,
                nextReviewAt,
                masterLevel,
            },
        }),
        prisma.reciteLog.create({
            data: {
                userId,
                planId,
                action,
                reviewDate: now,
            },
        }),
    ]);

    return toPlanItem(updated);
}

async function stats(userId: number): Promise<{
    total: number;
    todayCount: number;
    masteredCount: number;
    calendar: Record<string, number>;
}> {
    const now = new Date();
    const todayStr = formatDate(now);

    const [total, todayCount, masteredCount, logs] = await Promise.all([
        prisma.recitePlan.count({ where: { userId } }),
        prisma.reciteLog.count({
            where: { userId, reviewDate: { gte: startOfDay(now) } },
        }),
        prisma.recitePlan.count({
            where: { userId, masterLevel: { gte: EBBINGHAUS.length - 1 } },
        }),
        prisma.reciteLog.findMany({
            where: { userId },
            select: { reviewDate: true },
        }),
    ]);

    const calendar: Record<string, number> = {};
    for (const log of logs) {
        const key = formatDate(log.reviewDate);
        calendar[key] = (calendar[key] ?? 0) + 1;
    }

    return { total, todayCount, masteredCount, calendar };
}

function toPlanItem(p: {
    id: number;
    userId: number;
    quoteId?: number | null;
    customQuoteId?: number | null;
    currentStage: number;
    nextReviewAt: Date;
    lastReviewAt?: Date | null;
    masterLevel: number;
    createdAt: Date;
}): PlanItem {
    return {
        id: p.id,
        userId: p.userId,
        quoteId: p.quoteId ?? undefined,
        customQuoteId: p.customQuoteId ?? undefined,
        currentStage: p.currentStage,
        nextReviewAt: p.nextReviewAt.toISOString(),
        lastReviewAt: p.lastReviewAt?.toISOString(),
        masterLevel: p.masterLevel,
        createdAt: p.createdAt.toISOString(),
    };
}

function formatDate(d: Date): string {
    const y = d.getFullYear();
    const m = String(d.getMonth() + 1).padStart(2, '0');
    const day = String(d.getDate()).padStart(2, '0');
    return `${y}-${m}-${day}`;
}

function startOfDay(d: Date): Date {
    const x = new Date(d);
    x.setHours(0, 0, 0, 0);
    return x;
}

export { EBBINGHAUS, list, add, todayReview, review, stats };

File: server/app/services/recite_plan_service.py
"""背诵计划服务

对应 TypeScript 版 server/src/services/recite_plan_service.ts
"""
from datetime import datetime, timedelta

from sqlalchemy.orm import Session
from sqlalchemy import func

from ..deps import BusinessError
from ..models import RecitePlan, ReciteLog

EBBINGHAUS = [1, 2, 4, 7, 15, 30, 60]


def _format_date(d: datetime) -> str:
    return f"{d.year}-{d.month:02d}-{d.day:02d}"


def _start_of_day(d: datetime) -> datetime:
    return d.replace(hour=0, minute=0, second=0, microsecond=0)


def _to_plan_item(p: RecitePlan) -> dict:
    return {
        "id": p.id,
        "userId": p.userId,
        "quoteId": p.quoteId or None,
        "customQuoteId": p.customQuoteId or None,
        "currentStage": p.currentStage,
        "nextReviewAt": p.nextReviewAt.isoformat() if p.nextReviewAt else None,
        "lastReviewAt": p.lastReviewAt.isoformat() if p.lastReviewAt else None,
        "masterLevel": p.masterLevel,
        "createdAt": p.createdAt.isoformat() if p.createdAt else None,
    }


def list(db: Session, userId: int) -> list:
    plans = (
        db.query(RecitePlan)
        .filter(RecitePlan.userId == userId)
        .order_by(RecitePlan.nextReviewAt.asc())
        .all()
    )
    return [_to_plan_item(p) for p in plans]


def add(db: Session, userId: int, input_data: dict) -> dict:
    quote_id = input_data.get("quoteId")
    custom_quote_id = input_data.get("customQuoteId")

    if not quote_id and not custom_quote_id:
        raise BusinessError(400, "quoteId 或 customQuoteId 不能同时为空")

    existing = db.query(RecitePlan).filter(
        RecitePlan.userId == userId,
        RecitePlan.quoteId == quote_id if quote_id else RecitePlan.quoteId.is_(None),
        RecitePlan.customQuoteId == custom_quote_id if custom_quote_id else RecitePlan.customQuoteId.is_(None),
    ).first()

    if existing:
        return _to_plan_item(existing)

    next_review_at = datetime.now() + timedelta(days=EBBINGHAUS[0])

    plan = RecitePlan(
        userId=userId,
        quoteId=quote_id,
        customQuoteId=custom_quote_id,
        currentStage=0,
        nextReviewAt=next_review_at,
    )
    db.add(plan)
    db.commit()
    db.refresh(plan)

    return _to_plan_item(plan)


def todayReview(db: Session, userId: int) -> list:
    now = datetime.now()
    plans = (
        db.query(RecitePlan)
        .filter(RecitePlan.userId == userId, RecitePlan.nextReviewAt <= now)
        .order_by(RecitePlan.nextReviewAt.asc())
        .all()
    )
    return [_to_plan_item(p) for p in plans]


def review(db: Session, planId: int, userId: int, action: str) -> dict:
    plan = db.query(RecitePlan).filter(RecitePlan.id == planId).first()
    if not plan:
        raise BusinessError(404, "背诵计划不存在", 404)
    if plan.userId != userId:
        raise BusinessError(403, "无权操作", 403)

    new_stage = plan.currentStage
    master_level = plan.masterLevel

    if action == "master":
        new_stage = min(plan.currentStage + 1, len(EBBINGHAUS) - 1)
        master_level = min(master_level + 1, len(EBBINGHAUS))
    elif action == "not_skilled":
        new_stage = max(plan.currentStage - 1, 0)

    now = datetime.now()
    days = EBBINGHAUS[new_stage] if new_stage < len(EBBINGHAUS) else EBBINGHAUS[-1]
    next_review_at = now + timedelta(days=days)

    try:
        plan.currentStage = new_stage
        plan.lastReviewAt = now
        plan.nextReviewAt = next_review_at
        plan.masterLevel = master_level

        log = ReciteLog(
            userId=userId,
            planId=planId,
            action=action,
            reviewDate=now,
        )
        db.add(log)
        db.commit()
        db.refresh(plan)
    except Exception:
        db.rollback()
        raise

    return _to_plan_item(plan)


def stats(db: Session, userId: int) -> dict:
    now = datetime.now()
    today_start = _start_of_day(now)

    total = db.query(RecitePlan).filter(RecitePlan.userId == userId).count()

    today_count = (
        db.query(ReciteLog)
        .filter(ReciteLog.userId == userId, ReciteLog.reviewDate >= today_start)
        .count()
    )

    mastered_count = (
        db.query(RecitePlan)
        .filter(RecitePlan.userId == userId, RecitePlan.masterLevel >= len(EBBINGHAUS) - 1)
        .count()
    )

    logs = db.query(ReciteLog.reviewDate).filter(ReciteLog.userId == userId).all()

    calendar: dict[str, int] = {}
    for log in logs:
        key = _format_date(log[0])
        calendar[key] = calendar.get(key, 0) + 1

    return {
        "total": total,
        "todayCount": today_count,
        "masteredCount": mastered_count,
        "calendar": calendar,
    }
